### ISSUE TITLE
Musing: Unify Map/Filter/FlatMap/Negate/etc

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#6361]
We have a large number of operators that each have some sane interpretation on their own, but seem to ultimately boil down to an instance of the same core Differential Dataflow operator:
```rust
    /// Joins each record against a collection defined by the function `logic`.
    ///
    /// This method performs what is essentially a join with the collection of records `(x, logic(x))`.
    /// Rather than materialize this second relation, `logic` is applied to each record and the appropriate
    /// modifications made to the results, namely joining timestamps and multiplying differences.
    pub fn join_function<D2, R2, I, L>(&self, mut logic: L) -> Collection<G, D2, <R2 as Multiply<R>>::Output>
    where G::Timestamp: Lattice,
          D2: Data,
          R2: Semigroup+Multiply<R>,
          <R2 as Multiply<R>>::Output: Data+Semigroup,
          I: IntoIterator<Item=(D2,G::Timestamp,R2)>,
          L: FnMut(D)->I+'static,
    {
        self.inner
            .flat_map(move |(x, t, d)| logic(x).into_iter().map(move |(x,t2,d2)| (x, t.join(&t2), d2.multiply(&d))))
            .as_collection()
    }
```
This is a function that acts on a collection with a single argument `logic: L` whose type is
```rust
    where
          I: IntoIterator<Item=(D2,G::Timestamp,R2)>,
          L: FnMut(D)->I+'static,
```
meaning that for each input `datum: D` it produces a *collection*: a multiset of records whose contents vary as a function of time. The behavior of the operator is to `join` the input collection against the collection defined by `(x, logic(x))` for all `x`.

In this framing, 
1. `Map` is a join against a static singleton collection containing the columns we should append as a function of `x`,
2. `Filter` is a join against a collection with no columns that contains either no records (`false`) or one record (`true`).
3. `FlatMap` is a join against the output of the flatmap functions, which in MZ at the moment are a list of `(data, diff)`, without a time component. So, that but with time = 0.
4. `Negate` is a join against the collection containing one no-column entry with weight -1.
5. Temporal filters, not really part of `Filter`, is a join between the collection containing `((), lower_bound(x), +1)` and `((), upper_bound(x), -1)` where the upper bound is allowed to not exist if the constraint doesn't exist.

But, all of these seem to boil down to the same DD operator, which has a super easy implementation (above). We could at some point pile all of these operators into the same framing, which might allow us to think more clearly about fusing them, moving them around, and pushing their implementations into various other operators.

**Comment by ruchirK** [comment]
little bit sleepy so this might be a garbage thought but does `Project` fit into this framing as well?

**Comment by frankmcsherry** [comment]
Not exactly, though everywhere that I think of writing a `join` I think of following it with a `Project` anyhow. Like, no point in forcing the person to produce all the input columns as well, so might as well plan to do a projection with each operator. It technically fits in the `join_function` framing because `logic` just indicates what data to produce as output, rather than actually populating the collection with data.

In the `MapFilterProject` work, all of the evaluation happens first, and then projection is the very last thing to happen (after e.g. temporal filters). Unlike all the other things above, the projection doesn't really commute, afaict.

But, it makes sense to have project hang out as well, just probably not the same type of thing exactly. Again, afaict.
